File: crypto.ts

export const PUBLIC_KEY_LENGTH = 32
export const PRIVATE_KEY_LENGTH = 32
export const SIGNATURE_LENGTH = 64

export const MAIN_NET_CHAIN_ID = 87 //W
export const TEST_NET_CHAIN_ID = 84 //T

/* Type aliases used to increase flexibility and be able
   to extend these types later on. Also type aliases allows
   names to be more self explanatory like in BASE58 case. */

export type TBytes = Uint8Array

export type TBase64 = string
export type TBase58 = string
export type TBase16 = string //Same as HEX

export type TChainId = string | number

export interface IAddress {
  address: TBytes
  toString(): TBase58
  fromPublicKey: (chainId: TChainId, publicKey: IPublicKey) => IKeyPair
  isValid: () => boolean
}
export interface IPublicKey {
  publicKey: TBytes
  toString(): TBase58
  address: IAddress
  isValid: () => boolean
  verify: (bytes: TBytes | TBase58, signature: TBytes | TBase58) => boolean
}
export interface IPrivateKey {
  privateKey: TBytes
  toString(): TBase58
  publicKey(): () => IPublicKey
  isValid: () => boolean
  signBytes<T extends TBytes | TBase58>: (bytes: T) => T
}
export interface IKeyPair {
  privateKey: IPrivateKey
  publicKey: IPublicKey
  fromSeed: (seed: string, nonce: number) => IKeyPair
  randomSeedPhrase: () => string
}

/* Waves Crypto is a collection of essential cryptography and hashing
   algorithms used by Waves, protocol entities and binary structures. */

export interface IWavesCrypto {
  //Hashing 
  blake2b: (input: TBytes) => TBytes
  keccak: (input: TBytes) => TBytes
  sha256: (input: TBytes) => TBytes

  //Base encoding\decoding
  base64Encode: (input: TBytes) => TBase64
  base64Decode: (input: TBase64) => TBytes //throws (invalid input)
  base58Encode: (input: TBytes) => TBase58
  base58Decode: (input: TBase58) => TBytes //throws (invalid input)
  base16Encode: (input: TBytes) => TBase16
  base16Decode: (input: TBase16) => TBytes //throws (invalid input)

  //Random
  randomBytes: (size: number) => TBytes
  randomSeedPhrase: () => string

  //Bytes hashing and signing
  signBytes<T extends TBytes | TBase58>: (bytes: T, privateKey: T) => T

  //Verification
  verifySignature<T extends TBytes | TBase58>: (bytes: T, signature: T, publicKey: T) => boolean
  verifyPublicKey<T extends TBytes | TBase58>: (publicKey: T) => boolean
  verifyAddress<T extends TBytes | TBase58>: (address: T, chainId: TChainId, publicKey: T) => boolean

  //TODO Messaging
  //sharedKey<T extends TBytes | TBase58>: (privateKeyFrom: T, publicKeyTo: T) => T
  //messageDecrypt<T extends TBytes | TBase58>: (sharedKey: T, encryptedMessage: T) => T
  //messageEncrypt<T extends TBytes | TBase58>: (sharedKey: T, message: T) => T

}

File: crypto.go
package wavesplatform

// Bytes is a type alias for the slice of bytes.
type Bytes []byte

// PublicKey is a string representation of a public key bytes in form of BASE58 string.
type PublicKey string

// PrivateKey is a string representation of a private key in form of BASE58 string.
type PrivateKey string

// Seed is a BIP39 seed phrase.
type Seed string

// Address is a string representation of Waves address in form of BASE58 string.
type Address string

// KeyPair is an interface to a structure that holds corresponding private and public keys.
type KeyPair interface {
	PublicKey() string
	PrivateKey() string
}

// WavesChainID is a byte to represent blockchain identification.
type WavesChainID byte

// Known chain IDs
const (
	MainNet WavesChainID = 'W'
	TestNet WavesChainID = 'T'
)

// The lengths of basic crypto primitives.
const (
	PublicKeyLength  = 32
	PrivateKyeLength = 32
	SignatureLength  = 64
)

// WavesCrypto is a collection of functions to work with Waves basic types and crypto primitives used by Waves.
type WavesCrypto interface {
	Blake2b(input Bytes) Bytes // Blake2b produces the BLAKE2b-256 digest of the given `input`.
	Keccak(input Bytes) Bytes  // Keccak creates a new legacy Keccak-256 hash digest of the `input`.
	Sha256(input Bytes) Bytes  // Sha256 returns a new SHA256 checksum calculated from the `input`.

	Base58Encode(input Bytes) string // Base58Encode encodes the `input` into a BASE58 string.
	Base58Decode(input string) Bytes // Base58Decode decodes the `input` string to bytes.
	Base64Encode(input Bytes) string // Base64Encode returns a BASE64 string representation of the `input` bytes.
	Base64Decode(input string) Bytes // Base64Decode decodes the `input` BASE64 string to bytes.

	KeyPair(seed Seed) KeyPair       // KeyPair returns a pair of keys produced from the `seed`.
	PublicKey(seed Seed) PublicKey   // PublicKey returns a public key generated from the `seed`.
	PrivateKey(seed Seed) PrivateKey // PrivateKey generates a private key from the given `seed`.

	Address(publicKey PublicKey, chainID WavesChainID) Address // Address generates new Waves address from the `publicKey` and `chainID`.
	AddressFromSeed(seed Seed, chainID WavesChainID) Address   // AddressFromSeed returns a new Waves address produced from the `seed` and `chainID`.

	RandomSeed() Seed // RandomSeed return a new randomly generated BIP39 seed phrase.

	SignBytes(bytes Bytes, privateKey PrivateKey) Bytes // SignBytes produces a signature for the `bytes` by `privateKey`.
	SignBytesBySeed(bytes Bytes, seed Seed) Bytes       // SignBytesBySeed returns a signature for the `bytes` by a private keys generated from the `seed`.~``

	VerifySignature(publicKey PublicKey, bytes, signature Bytes) bool              // VerifySignature returns true if `signature` is a valid signature of `bytes` by `publicKey`.
	VerifyPublicKey(publicKey PublicKey) bool                                      // VerifyPublicKey returns true if `publicKey` is a valid public key.
	VerifyAddress(address Address, chainID WavesChainID, publicKey PublicKey) bool // VerifyAddress returns true if `address` is a valid Waves address for the given `chainId` and 	`publicKey`.
}
